src/ctapipe/image/modifications.py

```diff
@@ -100,6 +112,256 @@ def _smear_psf_randomly(
+    )
+
+    for i in range(n_noise_realizations):
+        if shuffle_full_cameras:
+            chosen = rng.permutation(n_events)[:nsb_level]
+            for event in chosen:
+                noise[i] += waveforms[event]
+        else:
+            for pixel in range(n_pixels):
+                chosen = rng.permutation(n_events)[:nsb_level]
+                # The line above is slower (especially for n_events much
+                # larger than nsb_level) than rng.choice(n_events, nsb_level)
+                # Unfortunately rng.choice does not currently work with numba.
+
+                for event in chosen:
+                    noise[i, :, pixel] += waveforms[event, :, pixel]
+    return noise
+
+
+class WaveformModifier(TelescopeComponent):
+    """
+    Component to add NSB noise to R1 waveforms.
+
+    This component in principle to be applied on MC simulations, to make them
+    closer to real data in terms of noise level.
+    The noise waveforms are read from a dedicated sim_telarray file, which
+    must be produced with the same telescope array configuration (and other
+    simulation settings) as the file to which the noise is to be added,
+    but containing only NSB noise (electronic noise should be switched off)
+
+    The number of simulated noise events per telescope in the NSB file must
+    be at least twice the number of waveforms ("nsb_level") from that file
+    that we want to add up. If the NSB file is produced with a level of 25%
+    of dark NSB, and we want to simulate 10x dark NSB, then nsb_level=40 (
+    =10/0.25) and the file must contain at least 80 events. This is to
+    guarantee that the different noise waveforms are not too correlated
+
+    This class can eventually use as NSB file a real data DL0 file which
+    contains pedestal events.
+
+    """
+
+    nsb_file = Path(
+        default_value=None,
+        help="Path to a dedicated NSB-only file (e.g. from sim_telarray)",
+    ).tag(config=True)
+
+    nsb_level = Int(
+        default_value=1,
+        help=(
+            "Number of random instances of the NSB waveforms from the "
+            "NSB file to be added up to the waveform"
+        ),
+    ).tag(config=True)
+
+    n_noise_realizations = Int(
+        default_value=100,
+        help=(
+            "Number of different realizations of the total NSB waveform to "
+            "be created per pixel"
+        ),
+    ).tag(config=True)
+
+    shuffle_full_cameras = Bool(
+        default_value=False,
+        help=(
+            "If True, full cameras are "
+            "combined to generate noise "
+            "waveforms for all pixels. Else, "
+            "each pixel uses a different random "
+            "combination of the inmput noise events"
+        ),
+    ).tag(config=True)
+
+    rng_seed = Int(default_value=1, help="Seed for the random number generator").tag(
+        config=True
+    )
+
+    def __init__(
```

Comment:
I think this `__init__` would benefit by being broken down into several smaller functions, e.g.:

```python

nsb_database = self.load_noise_events()
self.check_noise_statistics(nsb_database)
self.normalize_nsb_database(nsb_database)
self.nsb_database = self.stack_noise_waveforms(nsb_database)
```

Also, after the `total_noise` is computed, there is no need to keep the `nsb_database` around right?